kpatch-build/kpatch-elf.c

```diff
@@ -328,9 +341,29 @@ void kpatch_create_symbol_list(struct kpatch_elf *kelf)
 					sym->name);
 
 			if (is_bundleable(sym)) {
-				if (sym->sym.st_value != 0)
-					ERROR("symbol %s at offset %lu within section %s, expected 0",
+				/*
+				 * function prologue generated by gcc6 on ppc64le has
+				 * the sequence:
+				 *
+				 *	.globl my_func
+				 *	.type my_func, @function
+				 *	.quad .TOC.-my_func
+				 * my_func:
+				 *	.reloc ., R_PPC64_ENTRY ; optional
+				 *	ld r2,-8(r12)
+				 *	add r2,r2,r12
+				 * 	.localentry my_func, .-my_func
+				 *
+				 * where .localentry for the functions, will generated
+				 * st_value of 0x8 for the local functions. Skip, this
+				 * case.
+				 */
+				if (sym->sym.st_value != 0) {
+					if (!gcc6_check() ||
+					   (sym->type != STT_FUNC || sym->sym.st_value != 0x8))
+						ERROR("symbol %s at offset %lu within section %s, expected 0",
 					      sym->name, sym->sym.st_value, sym->sec->name);
+				}
```

Comment:
So I'm having trouble understanding this code.  It doesn't seem to match the comment.  It doesn't look like the symbol is skipped.  Instead, it's just ignored for GCC 6, and is otherwise reported as an error.  Is this code even needed?

Also I'm worried this check might produce some false positives.  Is there a more specific way to test for localentry symbols?

Looking at the binutils source for some clues:
```
  if (psym->st_other ^ ELF_ST_VISIBILITY (psym->st_other))
    printf (" [%s] ", get_symbol_other (psym->st_other ^ ELF_ST_VISIBILITY (psym->st_other)));

...

static const char *
get_ppc64_symbol_other (unsigned int other)
{
  if (PPC64_LOCAL_ENTRY_OFFSET (other) != 0)
    {
      static char buf[32];
      snprintf (buf, sizeof buf, _("<localentry>: %d"),
		PPC64_LOCAL_ENTRY_OFFSET (other));
      return buf;
    }
  return NULL;
}

...

#define PPC64_LOCAL_ENTRY_OFFSET(other)				\
  ppc64_decode_local_entry (((other) & STO_PPC64_LOCAL_MASK)	\
			    >> STO_PPC64_LOCAL_BIT)

...

/* The ELFv2 ABI uses three bits in the symbol st_other field of a
   function definition to specify the number of instructions between a
   function's global entry point and local entry point.
   The global entry point is used when it is necessary to set up the
   toc pointer (r2) for the function.  Callers must enter the global
   entry point with r12 set to the global entry point address.  On
   return from the function, r2 may have a different value to that
   which it had on entry.
   The local entry point is used when r2 is known to already be valid
   for the function.  There is no requirement on r12 when using the
   local entry point, and on return r2 will contain the same value as
   at entry.
   A value of zero in these bits means that the function has a single
   entry point with no requirement on r12 or r2, and that on return r2
   will contain the same value as at entry.
   Values of one and seven are reserved.  */
#define STO_PPC64_LOCAL_BIT		5
#define STO_PPC64_LOCAL_MASK		(7 << STO_PPC64_LOCAL_BIT)
```

So it looks like a specific check of the st_other field could be done to avoid false positives.  And then I think you could get rid of the gcc 6 check, which was problematic anyway for the case where kpatch-build was compiled with a different compiler as the kernel.

Or maybe this check isn't needed at all and this whole change can just be removed?